### ISSUE TITLE
Producer does not send message to kafka queue on linux

### ISSUE THREAD

**Issue by niraj-dodal** [confluentinc/confluent-kafka-dotnet#500]
Description
===========
We are trying to connect to Kafka Queue from the linux container in GCP environment. The producer does not send any message on the specified topic to the Kafka Queue. It kind of hangs at produceAsync Method.

we have received a console log something like below once -
"Delivered 'hi' to: @Invalid[-1001]"

This works fine on my local windows machine. Below is the producer code which we are using.
Any help around this is appreciated

How to reproduce
================

class SimpleProducer : ISimpleProducer
    {
        public void Produce(string message)
        {
            var config = new Dictionary<string, object>
            {
                {"bootstrap.servers", "localhost:9092" }
            };
            try
            {
                using (var producer = new Producer<Null, string>(config, null, new StringSerializer(Encoding.UTF8)))
                {
                    var result = producer.ProduceAsync("simpletest", null, message).Result;
                    Console.WriteLine($"Delivered '{dr.Value}' to: {dr.TopicPartitionOffset}");
                    producer.Flush(100);
                };
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ". Inner Exception - " + ex.InnerException.Message);
            }
     }
}

Checklist
=========
Please provide the following information:

 - [ ] Confluent.Kafka nuget version: v0.11.4
 - [ ] Apache Kafka version: 1.0.0
 - [ ] Client configuration: 
 - [ ] Operating system: Linux 
 - [ ] Provide logs (with "debug" : "..." as necessary in configuration)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by mhowlett** [comment]
it sounds like the producer isn't able to connect to the kafka cluster. try setting `debug` to `all` and looking at the logs. maybe also try using the console producer that comes with kafka to verify your connection settings are correct.

**Comment by Yurst** [comment]
I am having a similar issue producing messages from Confluent.Kafka (nuget version: v0.11.4) to Kafka (version 1.0.0) running on Linux (Ubuntu 16) in a docker container (confluent/cp-kafka, confluent version 4.0.0).  My producer config is as follows:
>             var config = new Dictionary<string, object>()
>             {
>                 ["bootstrap.servers"] = brokerList,
>                 ["client.id"] = "icf",
>                 ["retries"] = 0,
>                 ["batch.num.messages"] = 1,
>                 ["socket.blocking.max.ms"] = 1,
>                 ["socket.max.fails"] = 0,
>                 ["message.send.max.retries"] = 0,
>                 ["queue.buffering.max.ms"] = 1,
>                 ["default.topic.config"] = new Dictionary<string, object>
>                 {
>                     { "request.required.acks", 1 },
>                     { "request.timeout.ms", 5000 },
>                     { "message.timeout.ms", 10000 },
>                 },
>                 ["debug"] = "all",
>                 ["api.version.request"] = true,
>             };
>
>             var deliveryReport = producer.ProduceAsync(topicName, null, JsonStr);
> 
>             deliveryReport.ContinueWith(task =>
>             {
>                 if (task.Result.Error.HasError)
>                     Console.WriteLine($"IS ERROR RESULT {deliveryReport.Result.Error.Message}");
>                 else
>                     Console.WriteLine(string.Format("Delivered {0}\nPartition: {0}, Offset: {1}", task.Result.Value, task.Result.Partition, task.Result.Offset));
> 
>                 if (task.IsFaulted)
>                     Console.WriteLine("IS FAULTED");
> 
>                 if (task.Exception != null)
>                     Console.WriteLine(deliveryReport.Exception.Message);
> 
>                 if (task.IsCanceled)
>                     Console.WriteLine("IS CANCELLED");
>             });

The client is apparently connecting to the target Kafka cluster because the topic is created when the first message is produced.  However, all produced messages are resulting in "IS ERROR RESULT **Local: Message timed out**" after 10 seconds, and no messages are appearing in the Kafka topic.  The client debug "all" logs at startup are as follows.

> 7|2018-05-21 15:29:52.119|icf#producer-1|BROKER| 10.7.1.111:29092/bootstrap: Added new broker with NodeId -1
> 7|2018-05-21 15:29:52.119|icf#producer-1|BRKMAIN| :0/internal: Enter main broker thread
> 7|2018-05-21 15:29:52.119|icf#producer-1|STATE| :0/internal: Broker changed state INIT -> UP
> 7|2018-05-21 15:29:52.119|icf#producer-1|BROADCAST| Broadcasting state change
> 7|2018-05-21 15:29:52.119|icf#producer-1|BRKMAIN| 10.7.1.111:29092/bootstrap: Enter main broker thread
> 7|2018-05-21 15:29:52.119|icf#producer-1|CONNECT| 10.7.1.111:29092/bootstrap: broker in state INIT connecting
> 7|2018-05-21 15:29:52.119|icf#producer-1|CONNECT| 10.7.1.111:29092/bootstrap: Connecting to ipv4#10.7.1.111:29092 (plaintext) with socket 1544
> 7|2018-05-21 15:29:52.119|icf#producer-1|STATE| 10.7.1.111:29092/bootstrap: Broker changed state INIT -> CONNECT
> 7|2018-05-21 15:29:52.119|icf#producer-1|BROADCAST| Broadcasting state change
> 7|2018-05-21 15:29:52.135|icf#producer-1|CONNECT| 10.7.1.111:29092/bootstrap: Connected to ipv4#10.7.1.111:29092
> 7|2018-05-21 15:29:52.135|icf#producer-1|CONNECTED| 10.7.1.111:29092/bootstrap: Connected (#1)
> 7|2018-05-21 15:29:52.135|icf#producer-1|FEATURE| 10.7.1.111:29092/bootstrap: Updated enabled protocol features +ApiVersion to ApiVersion
> 7|2018-05-21 15:29:52.135|icf#producer-1|STATE| 10.7.1.111:29092/bootstrap: Broker changed state CONNECT -> APIVERSION_QUERY
> 7|2018-05-21 15:29:52.135|icf#producer-1|BROADCAST| Broadcasting state change
> 7|2018-05-21 15:29:52.150|icf#producer-1|SEND| 10.7.1.111:29092/bootstrap: Sent ApiVersionRequest (v0, 21 bytes @ 0, CorrId 1)

No further debug logs appear when the messages are being produced, and nothing relevant appears in the broker logs (nothing above INFO level).

Any insights on how I can further trouble-shoot this will be much appreciated. Thanks.

**Comment by mhowlett** [comment]
@Yurst - does your application exit immediately after the produce request? if so are you calling `Flush`? `Producer` `Dispose` will not wait for messages to be delivered.

**Comment by Yurst** [comment]
Thanks for the reply.  Yes, I am calling Flush() before exiting, and all tasks are completing with the "timed out" error after the 10-second message.timeout, while waiting on the flush().  I'm thinking the issue may have something to do with the broker being in a state of "APIVERSION_QUERY", or some unseen version-incompatibility response, but I'm at a loss.

**Comment by Yurst** [comment]
Ok, I got it working...  For one thing, I had an older version of the Confluent.Kafka.dll file (0.9.2-preview) in my project (which I first created over a year ago).  I thought it had been updated when I ran the NuGet command, "Update-Package Confluent.Kafka -Version 0.11.4", but apparently it had not.

The other thing that fixed it was: in the Kafka docker environment on the Linux side, I needed to set "KAFKA_ADVERTISED_LISTENERS" to its accessible IP address, rather than to "localhost".  Otherwise, the Windows client was looking for the Zookeeper broker on its own localhost.

Thanks again for your help.
